src/nvim/eval/typval.c

```diff
@@ -0,0 +1,2560 @@
+  FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC
+{
+  return xmalloc(sizeof(listitem_T));
+}
+
+/// Free a list item
+///
+/// Also clears the value. Does not touch watchers.
+///
+/// @param[out]  item  Item to free.
+void tv_list_item_free(listitem_T *const item)
+  FUNC_ATTR_NONNULL_ALL
+{
+  tv_clear(&item->li_tv);
+  xfree(item);
+}
+
+/// Remove a list item from a List and free it
+///
+/// Also clears the value.
+///
+/// @param[out]  l  List to remove item from.
+/// @param[in,out]  item  Item to remove.
+void tv_list_item_remove(list_T *const l, listitem_T *const item)
+  FUNC_ATTR_NONNULL_ALL
+{
+  tv_list_remove_items(l, item, item);
+  tv_list_item_free(item);
+}
+
+//{{{2 List watchers
+
+/// Add a watcher to a list
+///
+/// @param[out]  l  List to add watcher to.
+/// @param[in]  lw  Watcher to add.
+void tv_list_watch_add(list_T *const l, listwatch_T *const lw)
+  FUNC_ATTR_NONNULL_ALL
+{
+  lw->lw_next = l->lv_watch;
+  l->lv_watch = lw;
+}
+
+/// Remove a watcher from a list
+///
+/// Does not give a warning if watcher was not found.
+///
+/// @param[out]  l  List to remove watcher from.
+/// @param[in]  lwrem  Watcher to remove.
+void tv_list_watch_remove(list_T *const l, listwatch_T *const lwrem)
+  FUNC_ATTR_NONNULL_ALL
+{
+  listwatch_T **lwp = &l->lv_watch;
+  for (listwatch_T *lw = l->lv_watch; lw != NULL; lw = lw->lw_next) {
+    if (lw == lwrem) {
+      *lwp = lw->lw_next;
+      break;
+    }
+    lwp = &lw->lw_next;
+  }
+}
+
+/// Advance watchers to the next item
+///
+/// Used just before removing an item from a list.
+///
+/// @param[out]  l  List from which item is removed.
+/// @param[in]  item  List item being removed.
+void tv_list_watch_fix(list_T *const l, const listitem_T *const item)
+  FUNC_ATTR_NONNULL_ALL
+{
+  for (listwatch_T *lw = l->lv_watch; lw != NULL; lw = lw->lw_next) {
+    if (lw->lw_item == item) {
+      lw->lw_item = item->li_next;
+    }
+  }
+}
+
+//{{{2 Alloc/free
+
+/// Allocate an empty list
+///
+/// Caller should take care of the reference count.
+///
+/// @return [allocated] new list.
+list_T *tv_list_alloc(void)
+  FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC
+{
+  list_T *const list = xcalloc(1, sizeof(list_T));
+
+  // Prepend the list to the list of lists for garbage collection.
+  if (gc_first_list != NULL) {
+    gc_first_list->lv_used_prev = list;
+  }
+  list->lv_used_prev = NULL;
+  list->lv_used_next = gc_first_list;
+  gc_first_list = list;
+  return list;
+}
+
+/// Free items contained in a list
+///
+/// @param[in,out]  l  List to clear.
+void tv_list_free_contents(list_T *const l)
+  FUNC_ATTR_NONNULL_ALL
+{
+  for (listitem_T *item = l->lv_first; item != NULL; item = l->lv_first) {
+    // Remove the item before deleting it.
+    l->lv_first = item->li_next;
+    tv_clear(&item->li_tv);
+    xfree(item);
+  }
+  l->lv_len = 0;
+  l->lv_idx_item = NULL;
+  for (listwatch_T *lw = l->lv_watch; lw != NULL; lw = lw->lw_next) {
+    lw->lw_item = NULL;
+  }
```

Comment:
They are supposed to be removed by the callers. If I am not mistaking, this may just be as well `assert(l->lv_watch == NULL)`: `free_for_info` which is the only user of `tv_list_watch_remove()` may only work with an allocated list and `tv_list_free_contents` is only called when freeing the list or when GCing it, in both cases it will be rendered unusable before `free_for_info` could do anything and lists are not supposed to be GCed if they are being referenced by `:for`.